### PR TITLE
Convert Engines To Requests

### PR DESCRIPTION
**Pull request by hank-bond**
Ticket: #12 
Type: Improvement

## Changes

- We are now using `requests` as a backend for `DefaultEngine` and children.
- Added tests for `DefaultEngine` and `TorEngine`

## Documentation

Tests are written with `pytest` but it looks like Travis is configured to work with `nose2`.  Do we convert the tests or the CI test env?

After merging this we can complete #8 and #5 

## Impacted Areas
- `TorEngine`
- `DefaultEngine`

## Checklist

- [x] Add information to the release notes documentation
- [x] Add new feature to the usage documentation
- [x] Add tests

### CODE REVIEW

File: rasp/base.py [hunk 1]
```diff
@@ -16,23 +13,27 @@ def cleanup(self):
 
 
 class DefaultEngine(Engine):
-    def __init__(self, data=None, headers=None):
-        self.data = data
+    def __init__(self, headers=None):
+        self.session = self._session()
         self.headers = headers or {'User-Agent': DEFAULT_USER_AGENT}
```
Comment (wdm0006) [inline rasp/base.py]: If these are really default headers, and not added to the headers passed directly to get_page_source (since you added headers there as well), then we should rename the instance variable to `default_headers`.   Otherwise, they should get merged in that method.

File: rasp/tor_engine.py [hunk 2]
```diff
@@ -1,57 +1,63 @@
...
-    def __init__(self, pw=None, control=None, signal=Signal.NEWNYM, proxy_handler=None, data=None, headers=None):
-        super(TorEngine, self).__init__(data, headers)
-
-        self.signal = signal
-        self.pw = pw or getpass.getpass("Tor password: ")
-        self.control = control or ("127.0.0.1", 9051)
-
-        default_handler = urllib.request.ProxyHandler({"http": "127.0.0.1:8118"})
-        self.proxy_handler = proxy_handler or default_handler
+    def __init__(self,
```
Comment (wdm0006) [inline rasp/tor_engine.py]: I like the option for environment variables here, but we need to document that.  Could you add a sphinx-style docstring explaining the order of defaults and what they are for these input params?

If you haven't written a docstring before it's something like: 

```
"""short description

Long description.

:param param1: description
...
:param N: description
"""
```

Comment (wdm0006) [reply rasp/tor_engine.py]: Note that this will get picked up by sphinx automatically when the docs are rebuilt, and put into the description for this class.

File: rasp/base.py (hunk 1 above)
Comment (wdm0006) [reply rasp/base.py]: Probably should add a doc string explaining the default nature of these as well.

**Review by wdm0006 — COMMENTED** [review]
Made a couple of comments on naming/docs, but the content of the code looks good to me.

File: rasp/tor_engine.py (hunk 2 above)
Comment (hank-bond) [reply rasp/tor_engine.py]: Sure, are we using the default format or another (imo better to read) format like the Numpy/Google docstring format?

File: rasp/base.py (hunk 1 above)
Comment (hank-bond) [reply rasp/base.py]: oh shoot, the plan was to have them merge. let me fix that.

File: rasp/tor_engine.py (hunk 2 above)
Comment (wdm0006) [reply rasp/tor_engine.py]: I did default in PR 18, but don't have a very strong opinion one way or another.  Currently the extension to do numpy style isn't in reqs or loaded, so it won't build, but it's not hard to add it.

**Comment by wdm0006** [comment]
Also it looks like the build is failing: https://travis-ci.org/anidata/rasp/jobs/210293342

**Comment by hank-bond** [comment]
Added docs and release note.  Ready to review @danlrobertson (i can't modify the PR reviewer field in this project)

**Review by wdm0006 — APPROVED** [review]
LGTM

File: rasp/tor_engine.py [hunk 3]
```diff
@@ -1,57 +1,109 @@
...
-        except urllib.error.HTTPError as e:
-            time.sleep(2)
+    def clean_circuits(self):
+        """Refreshes the mapping of nodes we connect through.
+
+        When a new path from the user to the destination is required, this
+        method refreshes the node path we use to connect. This gives us a new IP
+        address with which the destination sees."""
+        kwargs = {'address': self.address, 'port': self.control_port}
+        with Controller.from_port(**kwargs) as controller:
```
Comment (dlrobertson) [inline rasp/tor_engine.py]: `from_port` creates a queue and a bunch of other stuff. Create a `Controller` in the constructor and call `self.signal` here.

File: rasp/tor_engine.py [hunk 4]
```diff
@@ -1,57 +1,109 @@
...
+            controller.authenticate(password=self.control_password)
```
Comment (dlrobertson) [inline rasp/tor_engine.py]: `authenticate` can throw some exceptions. Most notably `AuthenticationError`. This is another benefit of calling this in the constructor.

File: .travis.yml [hunk 5]
```diff
@@ -7,6 +7,6 @@ install:
   - "pip install -r requirements-dev.txt"
   - "pip install coveralls"
 script:
-  - nose2
+  - py.test
 after_success:
   - coveralls
```
Comment (dlrobertson) [inline .travis.yml]: nit:can you add a newline at the end of this file

File: docs/source/release_notes.rst [hunk 6]
```diff
@@ -0,0 +1,8 @@
+========
+Releases
+========
+
+0.0.1
+=====
+
+- *Improved*: Switched from ``DefaultEngine`` using ``urllib`` to ``requests``
```
Comment (dlrobertson) [inline docs/source/release_notes.rst]: I'd rather stick to just detailed commits and commit messages

**Review by dlrobertson — CHANGES_REQUESTED** [review]
Mostly just nits. The only real critique has to do with `TorEngine.clean_circuits`. Unless there is a reason to create the `Controller` and authenticate in the `clean_circuits` function, I'd suggest creating it in the constructor.

File: rasp/tor_engine.py (hunk 4 above)
Comment (hank-bond) [reply rasp/tor_engine.py]: You bring up an excellent point that I did not consider.  We definitely should not be creating a new Controller object for each signal being sent.  Even if the cost of creation is low, it's still dumb suboptimal. 

I still feel uncomfortable relying on users to call a `cleanup` method when we can just have the Engine created within a context manager.  We need to do this for the `SeleniumEngine`, so I feel like we might as well make all the engines use the `with` keyword even if they don't require any opening/closing (like the `DefaultEngine`).

There are a few different ways in which me might encapsulate this behavior in a context manager, and they have different UX considerations.  I will try a few tomorrow and present the ones I think are clearest.

Comment (dlrobertson) [reply rasp/tor_engine.py]: > I still feel uncomfortable relying on users to call a cleanup method

I totally agree. IMO we should also impl the `__del__` destructor and hope that our users do one of 1) use the `with` syntax 2) don't cause a circular reference. If we impl both `__del__` and `__exit__` we'll have to make sure we don't cause a double free when `__exit__` is called and then the good ol' gc comes along.

Comment (wdm0006) [reply rasp/tor_engine.py]: I do like the context manager syntax a lot, but would caution not to make that the only way to use the engine.  A pretty straight forward pattern is to leave something like `cleanup` and call that in the `exit`, so that you can do either context manager style or the "normal" way.  Ideally in both cases the user does nothing to clean up, so I'd say `__del__` and `__exit__` both calling a `__cleanup()` or something like that would make most sense.

Also for the `__del__` - `__exit__` clash, my understanding is that `__del__` would only get called when the refcount dropped to zero regardless, so I don't think they practically can clash, unless you explicitly called `del foo` on the object after the context manager had closed.  I might be off on that though.

Comment (dlrobertson) [reply rasp/tor_engine.py]: I could be wrong here, but I was pretty sure that `__del__` was essentially the python equivalent of the `tp_dealloc` member of `PyTypeObject`. I thought that `__exit__` and `__del__` were not exclusive so that `__del__` was essentially called after `__exit__`. Now that I think about it is implementing `__del__` when implementing `__exit__` an anti-pattern? I've never gone the context manager route before, so this is all new to me :-)

> make sure we don't cause a double free 

Example demonstrating what I was trying to say.

```python
import gc

class Test(object):
    def __init__(self):
        self.i = 0

    def info(self):
        return (self.i, hex(id(self)))

    def __enter__(self):
        self

    def __exit__(self, *a):
        self.i += 1
        print("__exit__: i={} &self={}".format(*self.info()))

    def __del__(self):
        self.i += 1
        print("__del__: i={} &self={}".format(*self.info()))

first = Test()
assert len(gc.get_referrers(first)) == 1
with first:
    assert len(gc.get_referrers(first)) == 2
assert len(gc.get_referrers(first)) == 1

with Test() as second:
    pass
```
Example run
```
$ python test.py
__exit__: i=1 &self=0x7eff20c76198
__exit__: i=1 &self=0x7eff20c76208
__del__: i=2 &self=0x7eff20c76208
__del__: i=2 &self=0x7eff20c76198
```

Comment (wdm0006) [reply rasp/tor_engine.py]: Huh, I played around with your script (doesn't break anything here, but `__enter__` should actually return self).  

If you call `__del__`, in the `__exit__` it does end up getting called twice.  If you call `__exit__` in your `__del__` and have all of your cleanup logic there, then either mechanism works out.  Not really advocating that, but regardless, I guess you shouldn't call `__del__` from `__exit__`.

Comment (hank-bond) [reply rasp/tor_engine.py]: The problem with relying on `__del__` is that there's no promise when it will be called, and is hidden from the control flow of the program. The `__enter__` and `__exit__` of a context manager is explicit, and will ensure leaks won't happen.

@danlrobertson You're right that with CPython, the GC is keeping track of the refcount and when it hits 0 the object is destructed, but we can't guarantee that the user will be using CPython.  The refcount is also based on scope, and sometimes things stay in scope for longer than can be determined by the static control flow of the code (think attributes inside of a decorator). 

The biggest concern for me is that when an exception is raised while executing the controller signal, `__del__` straight up won't run.  This is the second major reason why we have context managers to begin with.  I think we should go with @wdm0006 suggestion and utilize both context managers, and the ability to close out the connection manually (which someone could do in a try/except clause).

Also, keep in mind that if we find ourselves with an error raised in `__init__`, the object won't complete initialization, thus won't ever have a non-zero refcount and will never have its `__del__` method called.

Basically, we shouldn't use `__del__`.

Comment (wdm0006) [reply rasp/tor_engine.py]: Sounds good to me.  So general principle here is 'never rely on `__del__`, but also don't force use of context managers (as much as they are the best option much of the time)'.  Fair?

**Review by dlrobertson — APPROVED** [review]
:+1:

**Comment by hank-bond** [comment]
For brevity's sake this is the usage.

```python
from rasp.tor_engine import TorEngine

engine = TorEngine(control_password='raspdefaulttorpass')
with engine.refreshable_ip():
    page = engine.get_page_source(
        'http://www.google.com', refresh_ip=True
    )
    print(page.source)
```

In a future PR the `refresh_ip` flag will most likely be moved to the `TorEngine.__init__` method.
